pkg/commands/root.go

```diff
@@ -1,160 +1,168 @@
+	// Otherwise, string slice options will be duplicated.
+	fs := pflag.NewFlagSet("config flag set", pflag.ContinueOnError)
 
-	fs.StringVar(&cfg.Run.CPUProfilePath, "cpu-profile-path", "", wh("Path to CPU profile output file"))
-	fs.StringVar(&cfg.Run.MemProfilePath, "mem-profile-path", "", wh("Path to memory profile output file"))
-	fs.StringVar(&cfg.Run.TracePath, "trace-path", "", wh("Path to trace output file"))
+	// Ignore unknown flags because we will parse the command flags later.
+	fs.ParseErrorsWhitelist = pflag.ParseErrorsWhitelist{UnknownFlags: true}
 
-	fs.IntVarP(&cfg.Run.Concurrency, "concurrency", "j", getDefaultConcurrency(),
-		wh("Number of CPUs to use (Default: number of logical CPUs)"))
+	opts := &rootOptions{}
 
-	fs.BoolVar(&cfg.Run.PrintVersion, "version", false, wh("Print version"))
-}
+	// Don't do `fs.AddFlagSet(cmd.Flags())` because it shares flags representations:
+	// `changed` variable inside string slice vars will be shared.
+	// Use another config variable here,
+	// to not affect main parsing by this parsing of only config option.
+	setupRootPersistentFlags(fs, opts)
 
-func printMemStats(ms *runtime.MemStats, logger logutils.Log) {
-	logger.Infof("Mem stats: alloc=%s total_alloc=%s sys=%s "+
-		"heap_alloc=%s heap_sys=%s heap_idle=%s heap_released=%s heap_in_use=%s "+
-		"stack_in_use=%s stack_sys=%s "+
-		"mspan_sys=%s mcache_sys=%s buck_hash_sys=%s gc_sys=%s other_sys=%s "+
-		"mallocs_n=%d frees_n=%d heap_objects_n=%d gc_cpu_fraction=%.2f",
-		formatMemory(ms.Alloc), formatMemory(ms.TotalAlloc), formatMemory(ms.Sys),
-		formatMemory(ms.HeapAlloc), formatMemory(ms.HeapSys),
-		formatMemory(ms.HeapIdle), formatMemory(ms.HeapReleased), formatMemory(ms.HeapInuse),
-		formatMemory(ms.StackInuse), formatMemory(ms.StackSys),
-		formatMemory(ms.MSpanSys), formatMemory(ms.MCacheSys), formatMemory(ms.BuckHashSys),
-		formatMemory(ms.GCSys), formatMemory(ms.OtherSys),
-		ms.Mallocs, ms.Frees, ms.HeapObjects, ms.GCCPUFraction)
-}
+	fs.Usage = func() {} // otherwise, help text will be printed twice
 
-func formatMemory(memBytes uint64) string {
-	const Kb = 1024
-	const Mb = Kb * 1024
+	if err := fs.Parse(safeArgs(fs, os.Args)); err != nil {
+		if errors.Is(err, pflag.ErrHelp) {
+			return nil, err
+		}
 
-	if memBytes < Kb {
-		return fmt.Sprintf("%db", memBytes)
+		return nil, fmt.Errorf("can't parse args: %w", err)
 	}
-	if memBytes < Mb {
-		return fmt.Sprintf("%dkb", memBytes/Kb)
-	}
-	return fmt.Sprintf("%dmb", memBytes/Mb)
+
+	return opts, nil
 }
 
-func getDefaultConcurrency() int {
-	if os.Getenv(envHelpRun) == "1" {
-		// Make stable concurrency for generating help documentation.
-		const prettyConcurrency = 8
-		return prettyConcurrency
+// Shorthands are a problem because pflag, with UnknownFlags, will try to parse all the letters as options.
+// A shorthand can aggregate several letters (ex `ps -aux`)
+// The function replaces non-supported shorthands by a dumb flag.
```

Comment:
Is this the same as the current behavior? What's the rationale behind silently ignoring such faulty usage?

Reply:
> Is this the same as the current behavior?

Yes, before the "pre-parsing" was done with all the flags but only some were used.

The function works in isolation, there are no impact on the effective flag parsing.